Quarkus project using hibernate validator on classes from separate jars does not compile with 1.13.5.Final

## Describe the bug
(Describe the problem clearly and concisely.)
I have a couple of projects that use DTO classes from another jar that are annotated with hibernate validator classes. This includes validations on type declarations for container classes. When I moved from 1.13.4.Final to 1.13.5.final, my project failed to compile.

### Expected behavior
(Describe the expected behavior clearly and concisely.)
I expected everything to compile as with past Quarkus version upgrades.

### Actual behavior
(Describe the actual behavior clearly and concisely.)
The compile for both project failed with stack trace:
F
[demo1.zip](https://github.com/quarkusio/quarkus/files/6550261/demo1.zip)
ailed to execute goal io.quarkus:quarkus-maven-plugin:1.13.5.Final:build (default) on project scoreapi: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] [error]: Build step io.quarkus.hibernate.validator.deployment.HibernateValidatorProcessor#build threw an exception: java.lang.ClassCastException: class org.jboss.jandex.ClassInfo cannot be cast to class org.jboss.jandex.MethodInfo (org.jboss.jandex.ClassInfo and org.jboss.jandex.MethodInfo are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @5411fed2)
[ERROR] at org.jboss.jandex.MethodParameterTypeTarget.enclosingTarget(MethodParameterTypeTarget.java:56)
[ERROR] at org.jboss.jandex.MethodParameterTypeTarget.enclosingTarget(MethodParameterTypeTarget.java:39)
[ERROR] at io.quarkus.hibernate.validator.deployment.HibernateValidatorProcessor.build(HibernateValidatorProcessor.java:281)
[ERROR] at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
[ERROR] at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
[ERROR] at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
[ERROR] at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] at org.jboss.threads.JBossThread.run(JBossThread.java:501)

## To Reproduce

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

Or attach an archive containing the reproducer to the issue.

Steps to reproduce the behavior:
1. reproducer attached.
2. 
3. 

### Configuration
```properties
# Add your application.properties here, if applicable.

```

### Screenshots
(If applicable, add screenshots to help explain your problem.)

## Environment (please complete the following information):

### Output of `uname -a` or `ver`
Darwin BillMacBookPro.local 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64
### Output of `java -version`
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 11.0.11+8-jvmci-21.1-b05, mixed mode, sharing)
### GraalVM version (if different from Java)
CE 21.1.0
### Quarkus version or git rev
1.13.4.Final and 1.13.5.Final(attempted)
### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /usr/local/Cellar/maven/3.8.1/libexec
Java version: 11.0.11, vendor: GraalVM Community, runtime: /Library/Java/JavaVirtualMachines/graalvm-ce-java11-21.1.0/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "11.4", arch: "x86_64", family: "mac"
## Additional context
(Add any other context about the problem here.)

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -46,7 +46,7 @@
         <smallrye-health.version>3.0.2</smallrye-health.version>
         <smallrye-metrics.version>3.0.1</smallrye-metrics.version>
         <smallrye-open-api.version>2.1.5</smallrye-open-api.version>
-        <smallrye-graphql.version>1.2.3</smallrye-graphql.version>
+        <smallrye-graphql.version>1.2.4</smallrye-graphql.version>
         <smallrye-opentracing.version>2.0.0</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>5.1.0</smallrye-fault-tolerance.version>
         <smallrye-jwt.version>3.2.0</smallrye-jwt.version>

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -17,7 +17,7 @@
         <bouncycastle.version>1.68</bouncycastle.version>
         <bouncycastle.fips.version>1.0.2.1</bouncycastle.fips.version>
         <bouncycastle.tls.fips.version>1.0.11.4</bouncycastle.tls.fips.version>
-        <jandex.version>2.2.3.Final</jandex.version>
+        <jandex.version>2.3.0.Final</jandex.version>
         <resteasy.version>4.6.0.Final</resteasy.version>
         <opentracing.version>0.33.0</opentracing.version>
         <opentracing-jaxrs.version>1.0.0</opentracing-jaxrs.version>

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -548,7 +548,7 @@
                 <plugin>
                     <groupId>org.jboss.jandex</groupId>
                     <artifactId>jandex-maven-plugin</artifactId>
-                    <version>1.0.8</version>
+                    <version>1.1.0</version>
                 </plugin>
                 <plugin>
                     <groupId>net.revelc.code.formatter</groupId>
@@ -912,7 +912,7 @@
                                                     jandex-maven-plugin
                                                 </artifactId>
                                                 <versionRange>
-                                                    [1.0.5,)
+                                                    [1.1.0,)
                                                 </versionRange>
                                                 <goals>
                                                     <goal>jandex</goal>

--- a/docs/src/main/asciidoc/cdi-reference.adoc
+++ b/docs/src/main/asciidoc/cdi-reference.adoc
@@ -53,7 +53,7 @@ To generate the index just add the following to your `pom.xml`:
     <plugin>
       <groupId>org.jboss.jandex</groupId>
       <artifactId>jandex-maven-plugin</artifactId>
-      <version>1.0.7</version>
+      <version>1.1.0</version>
       <executions>
         <execution>
           <id>make-index</id>

--- a/docs/src/main/asciidoc/cdi-reference.adoc
+++ b/docs/src/main/asciidoc/cdi-reference.adoc
@@ -72,7 +72,7 @@ If you are are using gradle, you can apply the following plugin to your `build.g
 [source,groovy]
 ----
 plugins {
-    id 'org.kordamp.gradle.jandex' version '0.10.0'
+    id 'org.kordamp.gradle.jandex' version '0.11.0'
 }
 ----
 

--- a/docs/src/main/asciidoc/maven-tooling.adoc
+++ b/docs/src/main/asciidoc/maven-tooling.adoc
@@ -551,7 +551,7 @@ unless it is the main application module already configured with the quarkus-mav
     <plugin>
       <groupId>org.jboss.jandex</groupId>
       <artifactId>jandex-maven-plugin</artifactId>
-      <version>1.0.7</version>
+      <version>1.1.0</version>
       <executions>
         <execution>
           <id>make-index</id>

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
@@ -273,37 +273,26 @@ class HibernateValidatorProcessor {
                     contributeClass(classNamesToBeValidated, indexView, annotation.target().asClass().name());
                     // no need for reflection in the case of a class level constraint
                 } else if (annotation.target().kind() == AnnotationTarget.Kind.TYPE) {
-                    // there is a bug in Jandex regarding enclosingTarget() when annotations are on constructors located in a separate indexed library
-                    // (enclosing target is a ClassInfo instead of a MethodInfo and enclosingTarget() tries to catch it to a MethodInfo
-                    // leading to a ClassCastException)
-                    // for now we swallow the exception, pending a bugfix in Jandex
-                    // see https://github.com/quarkusio/quarkus/issues/17491
-                    try {
-                        // container element constraints
-                        AnnotationTarget enclosingTarget = annotation.target().asType().enclosingTarget();
-                        if (enclosingTarget.kind() == AnnotationTarget.Kind.FIELD) {
-                            contributeClass(classNamesToBeValidated, indexView,
-                                    enclosingTarget.asField().declaringClass().name());
-                            reflectiveFields.produce(new ReflectiveFieldBuildItem(enclosingTarget.asField()));
-                            if (annotation.target().asType().target() != null) {
-                                contributeClassMarkedForCascadingValidation(classNamesToBeValidated, indexView,
-                                        consideredAnnotation,
-                                        annotation.target().asType().target());
-                            }
-                        } else if (enclosingTarget.kind() == AnnotationTarget.Kind.METHOD) {
-                            contributeClass(classNamesToBeValidated, indexView,
-                                    enclosingTarget.asMethod().declaringClass().name());
-                            reflectiveMethods.produce(new ReflectiveMethodBuildItem(enclosingTarget.asMethod()));
-                            if (annotation.target().asType().target() != null) {
-                                contributeClassMarkedForCascadingValidation(classNamesToBeValidated, indexView,
-                                        consideredAnnotation,
-                                        annotation.target().asType().target());
-                            }
-                            contributeMethodsWithInheritedValidation(methodsWithInheritedValidation, indexView,
-                                    enclosingTarget.asMethod());
+                    // container element constraints
+                    AnnotationTarget enclosingTarget = annotation.target().asType().enclosingTarget();
+                    if (enclosingTarget.kind() == AnnotationTarget.Kind.FIELD) {
+                        contributeClass(classNamesToBeValidated, indexView, enclosingTarget.asField().declaringClass().name());
+                        reflectiveFields.produce(new ReflectiveFieldBuildItem(enclosingTarget.asField()));
+                        if (annotation.target().asType().target() != null) {
+                            contributeClassMarkedForCascadingValidation(classNamesToBeValidated, indexView,
+                                    consideredAnnotation,
+                                    annotation.target().asType().target());
                         }
-                    } catch (Exception e) {
-                        // ignore
+                    } else if (enclosingTarget.kind() == AnnotationTarget.Kind.METHOD) {
+                        contributeClass(classNamesToBeValidated, indexView, enclosingTarget.asMethod().declaringClass().name());
+                        reflectiveMethods.produce(new ReflectiveMethodBuildItem(enclosingTarget.asMethod()));
+                        if (annotation.target().asType().target() != null) {
+                            contributeClassMarkedForCascadingValidation(classNamesToBeValidated, indexView,
+                                    consideredAnnotation,
+                                    annotation.target().asType().target());
+                        }
+                        contributeMethodsWithInheritedValidation(methodsWithInheritedValidation, indexView,
+                                enclosingTarget.asMethod());
                     }
                 }
             }

--- a/independent-projects/arc/pom.xml
+++ b/independent-projects/arc/pom.xml
@@ -38,7 +38,7 @@
         <!-- Versions -->
         <version.cdi>2.0.2</version.cdi>
         <version.jta>1.3.3</version.jta>
-        <version.jandex>2.2.3.Final</version.jandex>
+        <version.jandex>2.3.0.Final</version.jandex>
         <version.junit5>5.7.2</version.junit5>
         <version.maven>3.8.1</version.maven>
         <version.assertj>3.19.0</version.assertj>

--- a/independent-projects/resteasy-reactive/pom.xml
+++ b/independent-projects/resteasy-reactive/pom.xml
@@ -37,7 +37,7 @@
         <maven.compiler.release>11</maven.compiler.release>
         <!-- Versions -->
         <version.cdi>2.0.2</version.cdi>
-        <version.jandex>2.2.3.Final</version.jandex>
+        <version.jandex>2.3.0.Final</version.jandex>
         <version.junit5>5.7.2</version.junit5>
         <version.maven>3.8.1</version.maven>
         <version.assertj>3.19.0</version.assertj>

--- a/independent-projects/tools/artifact-api/pom.xml
+++ b/independent-projects/tools/artifact-api/pom.xml
@@ -18,7 +18,7 @@
             <plugin>
                 <groupId>org.jboss.jandex</groupId>
                 <artifactId>jandex-maven-plugin</artifactId>
-                <version>1.0.8</version>
+                <version>1.1.0</version>
                 <executions>
                     <execution>
                         <id>make-index</id>

--- a/independent-projects/tools/registry-client/pom.xml
+++ b/independent-projects/tools/registry-client/pom.xml
@@ -50,7 +50,7 @@
             <plugin>
                 <groupId>org.jboss.jandex</groupId>
                 <artifactId>jandex-maven-plugin</artifactId>
-                <version>1.0.8</version>
+                <version>1.1.0</version>
                 <executions>
                     <execution>
                         <id>make-index</id>

--- a/integration-tests/gradle/src/test/resources/jandex-basic-multi-module-project/common/build.gradle
+++ b/integration-tests/gradle/src/test/resources/jandex-basic-multi-module-project/common/build.gradle
@@ -1,7 +1,7 @@
 plugins {
     id 'io.quarkus'
     id 'java-library'
-    id "org.kordamp.gradle.jandex" version "0.10.0"
+    id "org.kordamp.gradle.jandex" version "0.11.0"
 }
 
 dependencies {

--- a/tcks/microprofile-rest-client-reactive/pom.xml
+++ b/tcks/microprofile-rest-client-reactive/pom.xml
@@ -88,7 +88,7 @@
             <plugin>
                 <groupId>org.jboss.jandex</groupId>
                 <artifactId>jandex-maven-plugin</artifactId>
-                <version>1.0.8</version>
+                <version>1.1.0</version>
                 <executions>
                     <execution>
                         <id>make-index</id>

